twake/backend/node/src/services/documents/services/index.ts

```diff
@@ -514,18 +516,30 @@ export class DocumentsService {
         throw Error("Drive item doesn't exist");
       }
 
-      if (
-        !item.access_info.authorized_entities.find(entity =>
-          entityIdentityCheck(entity.id, context),
-        )
-      ) {
-        return false;
+      if (token) {
+        if (!item.access_info.public.token) {
+          return false;
+        }
+
+        const { token: itemToken, level: itemLevel } = item.access_info.public;
+
+        return itemLevel === level && itemToken === token;
       }
 
       if (
-        item.access_info.unauthorized_entities.find(entity =>
-          entityIdentityCheck(entity.id, context),
-        )
+        !item.access_info.entities.find(entity => {
+          if (entity.level !== level) return false;
+
+          if (entity.type === "user" && entity.id === context.user.id) {
+            return true;
+          }
+
+          if (entity.type === "company" && entity.id === context.company.id) {
+            return true;
+          }
+
```

Comment:
Something is missing here, the "folder" case it looks like this:
```
if(entity.type === "folder" && entity.id !== this.ROOT && entity.id !== this.TRASH && await checkAccess(entity.id, ...)){
  return true;
}
```

To avoid starting with this expensive line, maybe do it as a last step or order the auth_entities to test this folder case in last position.